ports/stm32f4/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,403 @@
+    }
+    if(HAL_TIM_PWM_Start(&self->handle, self->channel) != HAL_OK) {
+        mp_raise_ValueError(translate("Error starting PWM"));
+    }
+
+    self->variable_frequency = variable_frequency;
+    self->frequency = frequency;
+    self->duty_cycle = duty;
+
+    return PWMOUT_OK;
+}
+
+bool common_hal_pulseio_pwmout_deinited(pulseio_pwmout_obj_t* self) {
+    return self->tim == mp_const_none;
+}
+
+void common_hal_pulseio_pwmout_deinit(pulseio_pwmout_obj_t* self) {
+    if (common_hal_pulseio_pwmout_deinited(self)) {
+        return;
+    }
+    //var freq shuts down entire timer, others just their channel
+    if(self->variable_frequency) {
+        reserved_tim[self->tim->tim_index-1] = 0x00;
+    } else {
+        reserved_tim[self->tim->tim_index-1] &= ~(1<<self->tim->channel_index);
+        HAL_TIM_PWM_Stop(&self->handle, self->channel);
+    }
+    reset_pin_number(self->tim->pin->port,self->tim->pin->number);
+    self->tim = mp_const_none;
+
+    //if reserved timer has no active channels, we can disable it
+    if (!reserved_tim[self->tim->tim_index-1]) {
+        tim_frequencies[self->tim->tim_index-1] = 0x00;
+        tim_clock_disable(1<<(self->tim->tim_index-1));
+    }
+}
+
+void common_hal_pulseio_pwmout_set_duty_cycle(pulseio_pwmout_obj_t* self, uint16_t duty_cycle) {
+    uint16_t duty = duty_cycle/655;
+    uint32_t period = PWM_MAX_FREQ/self->frequency;
+    uint32_t input = (period*duty)/100;
+    //TODO: remove before merging
+    //mp_printf(&mp_plat_print, "duty_cycle %d, Duty: %d, Input %d\n", duty_cycle, duty, input);
+    __HAL_TIM_SET_COMPARE(&self->handle, self->channel, input);
+
+    self->duty_cycle = duty;
+}
+
+uint16_t common_hal_pulseio_pwmout_get_duty_cycle(pulseio_pwmout_obj_t* self) {
+    return self->duty_cycle;
+}
+
+void common_hal_pulseio_pwmout_set_frequency(pulseio_pwmout_obj_t* self, uint32_t frequency) {
+    if (frequency == 0 || frequency > 6000000) {
+        mp_raise_ValueError(translate("Invalid PWM frequency"));
+    }
+    if (frequency == self->frequency) return;
+
+    uint32_t source_freq = timer_get_source_freq(self->tim->tim_index);
+    uint32_t period = PWM_MAX_FREQ/frequency;
+
+    //shut down
+    HAL_TIM_PWM_Stop(&self->handle, self->channel);
+
+    //Only change altered values
+    self->handle.Init.Period = period - 1;
+    self->handle.Init.Prescaler = (source_freq / PWM_MAX_FREQ) - 1; // TIM runs at ~6MHz
+
+    //restart everything, adjusting for new speed
+    if(HAL_TIM_PWM_Init(&self->handle) != HAL_OK) {
+        mp_raise_ValueError(translate("Timer Re-Init Error"));
+    }
+    self->chan_handle.Pulse = (period*self->duty_cycle)/100 - 1;
+    if(HAL_TIM_PWM_ConfigChannel(&self->handle, &self->chan_handle, self->channel) != HAL_OK) {
+        mp_raise_ValueError(translate("Channel Re-Init Error"));
+    }
+    if(HAL_TIM_PWM_Start(&self->handle, self->channel) != HAL_OK) {
+        mp_raise_ValueError(translate("Error restarting PWM"));
+    }
```

Comment:
These kinds of errors don't mean a lot to the end user, and each one needs to be translated. Could merge them as general PWMOut errors, maybe also including a %d error code if you want.

Reply:
Good idea. Do you have an example I could follow style-wise, by any chance?

Reply:
From `_bleio`, we have many things like this, where we print out an error code that was returned from the underlying HAL. In this case, we don't have that, but, the point is just to show you how to print a numeric value. This one uses %x for hex, but decimal makes more sense here.
```
        mp_raise_OSError_msg_varg(translate("Failed to read CCCD value, err 0x%04x"), err_code);
```
But I am rethinking this, and maybe the current errors are not really so obscure. I might phrase them as "Could not re-init timer", "Could not re-init timer channel", "Could not restart PWM". Not a big deal.